examples/fileupload/serviceUpload.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/graphql-go/graphql"
+	handler "github.com/jpascal/graphql-upload"
+	uuid "github.com/satori/go.uuid"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path"
+)
+
+var UploadType = graphql.NewScalar(graphql.ScalarConfig{
+	Name:        "Upload",
+	Description: "Scalar upload object",
+})
+
+type FileWrapper struct {
+	File *os.File
+	Name string
+}
+
+var File = graphql.NewObject(graphql.ObjectConfig{
+	Name:        "File",
+	Description: "File object",
+	Fields: graphql.Fields{
+		"name": &graphql.Field{
+			Type: graphql.String,
+			Resolve: func(params graphql.ResolveParams) (interface{}, error) {
+				file := params.Source.(*FileWrapper)
+				name := path.Base(file.Name)
+
+				return name, nil
+			},
+		},
+		"hash": &graphql.Field{
+			Type: graphql.String,
+			Resolve: func(params graphql.ResolveParams) (interface{}, error) {
+				file := params.Source.(*FileWrapper)
+				if data, err := ioutil.ReadAll(file.File); err == nil {
+					fileHash := sha1.Sum(data)
+
+					return hex.EncodeToString(fileHash[:]), nil
+				} else {
+					return nil, err
+				}
+
+			},
+		},
+		"size": &graphql.Field{
+			Type: graphql.Int,
+			Resolve: func(params graphql.ResolveParams) (interface{}, error) {
+				file := params.Source.(*FileWrapper)
+				if info, err := file.File.Stat(); err != nil {
+					return nil, err
+				} else {
+					return info.Size(), nil
+				}
+			},
+		},
+	},
+})
+
+func main() {
+	schema, err := graphql.NewSchema(
+		graphql.SchemaConfig{
+			Query: graphql.NewObject(
+				graphql.ObjectConfig{
+					Name: "Query",
+					Fields: graphql.Fields{
+						"file": &graphql.Field{
+							Type: File,
+							Args: graphql.FieldConfigArgument{
+								"id": &graphql.ArgumentConfig{
+									Type: graphql.NewNonNull(graphql.String),
+								},
+							},
+							Resolve: func(params graphql.ResolveParams) (interface{}, error) {
+								if fileId, ok := params.Args["id"].(string); ok {
+									fileUuid, err := uuid.FromString(fileId)
+									if err != nil {
+										return nil, err
+									}
+
+									file, err := os.Open("tmp/" + fileUuid.String())
+									if err != nil {
+										return nil, err
+									}
+
+									return &FileWrapper{File: file, Name: fileUuid.String()}, nil
+								} else {
+									return nil, errors.New("file id is not provided")
+								}
+							},
+						},
+					},
+				}),
+			Mutation: graphql.NewObject(
+				graphql.ObjectConfig{
+					Name: "Mutation",
+					Fields: graphql.Fields{
+						"upload": &graphql.Field{
+							Type: graphql.NewNonNull(graphql.String),
+							Args: graphql.FieldConfigArgument{
+								"file": &graphql.ArgumentConfig{
+									Type: graphql.NewNonNull(UploadType),
+								},
+							},
+							Resolve: func(params graphql.ResolveParams) (interface{}, error) {
+								upload, uploadPresent := params.Args["file"].(handler.File)
```

Comment:
Okay, I think I'm following this PR for the most part. Once thing that I am missing is why this field would be `handler.Field`. If shouldn't it be `graphql.Upload` since that's what `injectFiles` sets for the variables?

Reply:
This is not the part of the gateway change - this file is only used as an example and in this "example server" supporting file uploads neither nautilus/gateway nor nautilus/graphql is used. This example service could be even implemented just in JS but for convenience, I've chosen following libraries: github.com/graphql-go/graphql, github.com/jpascal/graphql-upload.  We need this service to demonstrate that nautilus gateway can proxy file uploads. 

`github.com/jpascal/graphql-upload` has its own abstractions for file uploads (which is handler.File) and graphql. Upload is only used within the gateway itself. I mean, they represent the same thing but in different codebases. I didn't find any way to use that library in nautilus gateway & graphql itself as it is not tailored for sending requests (only fits for handling) so extra type was needed at least for nautilus/graphql.  After I implemented it there I didn't see any point to use another library type in the gateway.

Reply:
Ah, yea I got confused by the file. Thanks for clarifying!